usdot-fhwa-stol/CARMAPlatform
# Configured CARMA Deployment in Docker

This adds Dockerfiles to create CARMA images for Docker as well as several shell scripts that are necessary or very useful in the process of utilizing Docker.

**Notable Files:**
* `Dockerfile` - Primary Docker image construction mechanism, specifies all dependencies and build steps to create a fully operational CARMA system from a baseline ROS install
* `docker/Dockerfile.clean_build` - Same as above except instead of copying the local CARMAPlatform files it clones them from Github every time. Use the `--build-arg CARMA_VERSION=<VERSION>` to specify the branch/tag/commit to checkout.
* `engineering_tools/install_docker.sh` - Installation script for docker-ce
* `engineering_tools/run_carma.sh` - Script to start CARMA after the docker image has been created. Primarily responsible for setting the `docker run` command line arguments to remap files and configure networking.
* `docker/entrypoint.sh` - Start script run inside the docker container
* `docker/install.sh` - Analog to remote_install, except local.

**Procedure to Test**
1. Run `install-docker.sh` on local machine
2. Log out and back in to allow new usergroup settings to take effect
3. cd into CARMAPlatform
4. Run: `docker build . -t carma --build-arg SSH_PRIVATE_KEY="$(cat ~/.ssh/id_rsa)" --build-arg EXTRA_PACKAGES=git@github.com:fhwa-saxton/CARMASensitive.git` (id_rsa will not be preserved in final image due to multi-stage build process, we're not publishing these images for now anyway)
5. Run `./engineering_tools/run-carma.sh <RELATIVE_PATH_TO_CONFIGURATION_FOLDER>` where the config folder is cadillac_config, etc.
6. You should see some sort of hash output as a "PID" of sorts for that container instance. To see inside that container run `docker attach carma`. You should see a new TTY containing CARMA running inside a TMUX instance. To detach and leave it running press `Ctrl+P, Ctrl+Q`, to kill it press `Ctrl+C`. To kill it while it's in the background run `docker kill carma`, it will print the name of the killed container if successful.
